src/pipes/erldns_pipeline.erl

```diff
@@ -0,0 +1,245 @@
+-type host() :: inet:ip_address() | inet:hostname().
+-doc "The underlying request transport protocol. All requests come either through UDP or TCP.".
+-type transport() :: tcp | udp.
+-doc "Options that can be passed and accumulated to the pipeline.".
+-type opts() :: #{
+    resolved := boolean(),
+    transport := transport(),
+    host := host(),
+    atom() => dynamic()
+}.
+-type return() :: dns:message() | {dns:message(), opts()} | {stop, dns:message()}.
+-type pipe() :: module() | fun((dns:message(), opts()) -> return()).
+-type pipeline() :: [fun((dns:message(), opts()) -> return())].
+-export_type([transport/0, host/0, pipe/0, opts/0, return/0]).
+
+-doc """
+Initialise the pipe handler, triggering side-effects and preparing any necessary metadata.
+
+This will be called during the pipeline initialisation phase, which should happen at application
+startup provided you added the pipeline to your application's supervision tree. This will be called
+only once during application startup and therefore it is an opportunity to do any necessary
+preparations that can reduce the amount of work at runtime and therefore improve performance.
+
+This callback can return `disabled`, and then the `c:call/2` callback won't be added to the
+pipeline.
+""".
+-callback prepare(opts()) -> disabled | opts().
+-doc """
+Trigger the pipeline at run-time.
+
+This callback can return
+- a possibly new `t:dns:message/0`;
+- a tuple containing a new `t:dns:message/0` and a new set of `t:opts/0`;
+- a tuple `{stop, t:dns:message/0}` tuple to stop the pipeline execution altogether.
+""".
+-callback call(dns:message(), opts()) ->
+    dns:message() | {dns:message(), opts()} | {stop, dns:message()}.
+-optional_callbacks([prepare/1]).
+
+-behaviour(gen_server).
+
+-export([call/2]).
+-export([start_link/0, init/1, handle_call/3, handle_cast/2, terminate/2]).
+-ifdef(TEST).
+-export([def_opts/0]).
+-else.
+-compile({inline, [def_opts/0]}).
+-endif.
+
+-define(DEFAULT_PACKET_PIPELINE, [
+    erldns_query_throttle,
+    erldns_packet_cache,
+    erldns_resolver,
+    erldns_packet_cache,
+    erldns_empty_verification
+]).
+
+-spec call(dns:message(), #{atom() => dynamic()}) -> dns:message().
+call(Msg, Opts) ->
+    {Pipeline, DefOpts} = get_pipeline(),
+    do_call(Msg, Pipeline, maps:merge(DefOpts, Opts)).
+
+-spec do_call(dns:message(), pipeline(), opts()) -> dns:message().
+do_call(Msg, [Pipe | Pipes], Opts) when is_function(Pipe, 2) ->
+    try Pipe(Msg, Opts) of
+        #dns_message{} = Msg1 ->
+            do_call(Msg1, Pipes, Opts);
+        {#dns_message{} = Msg1, Opts1} ->
+            do_call(Msg1, Pipes, Opts1);
+        {stop, #dns_message{} = Msg1} ->
+            Msg1;
+        Other ->
+            ?LOG_ERROR(#{
+                what => pipe_failed,
+                pipe => Pipe,
+                msg => Msg,
+                opts => Opts,
+                unexpected_return => Other
+            }),
+            ct:pal("Value ~p~n", [Other]),
```

Comment:
Is this ok to be left here?

Reply:
Wtf, I'm sorry, the linter should have warned about this!